### DISCUSSION TITLE
Multiple Phidget Voltage Out Devices

### DISCUSSION THREAD

**Discussion by DB-stoutheart** [artisan-roaster-scope/artisan#793] [Q&A] [answered]
I have two Phidget Voltage Out Devices, one is the OUT1000_0, the other is the OUT1002_0. I've worked out how to send them a specific voltage, but I cannot figure out how to get the slider to work. I tried `out(0,{},610238:2)` but it doesn't do anything.  If I send out(0,2,610238:2) for a specific voltage it works fine.  Is this by chance a characteristic of it being an OUT1000 vs the 1002?

Thank you as always!
Dustin

**Comment by MAKOMO** [comment]
The command you indicated should work. Could you attach your artisan-settings.aset file (renamed with extension .txt to let it drag-and-drop here)?

**Reply by DB-stoutheart** [reply]
Okay, I'm traveling but will get it to you when I can.  Additionally, how can I address two separate phidget out objects with one button. In other words, right now when I press ON it triggers the gas with range(0,5). But I'd also like ON to trigger the fan to a specific setting. Does that make sense?

**Reply by MAKOMO** [reply]
You can put a sequence of commands separated by semicolons

**Comment by DB-stoutheart** [comment, marked as answer]
here you go, thank you for reviewing. 
[artisan-settings.txt](https://github.com/artisan-roaster-scope/artisan/files/7919171/artisan-settings.txt)

**Reply by MAKOMO** [reply]
Your air slider has a range of 0-100 and those values are 1:1 bound to {} but the OUT1000 has only a range from 0-4.2. You need to provide a corresponding factor here, right?

**Reply by DB-stoutheart** [reply]
I forgot about the Factor, and how to calculate it. Thank you for reminded me.

**Reply by MAKOMO** [reply]
I just added a Slider Mapping Calculator that calculates the Factor and Offset from your target min and max values. Check out the current continuous build.

The example shows the calculated results for the target min/max values of 0 to 4.2V.

<img width="1088" alt="Screenshot 2022-01-23 at 21 27 57" src="https://user-images.githubusercontent.com/13909011/150698112-620056a4-f2c5-4623-8c67-ebb165a3e670.png">

**Comment by DB-stoutheart** [comment]
Just ran the update, WOW...that Slider Mapping Calculator is a life changer!!!  Thank you Marko!

**Comment by DB-stoutheart** [comment]
One thing I can't seem to find any documentation on, I'm sure it exists but I can't find it, is how to get an LCD and graph the fan output.  Can you guide me through that or send me in the right direction?  Lastly, is there any way for the slider to always replicate whatever the current settings are?  So right now when I press "ON" in artisan, I have it so the fan goes to a specific setting, but the slider still says "0".

Thank you!

**Reply by MAKOMO** [reply]
The Air slider generates Air events. You need to make sure that you have those visible (the Air event needs to be ticked in the first tab of the Events dialog to have it drawn).

I don't understand your second question. How do you "have it so the fan goes to a specific setting" on ON?

**Comment by DB-stoutheart** [comment]
Regarding second question - I have the Events programmed so that when I click the "ON" button in Artisan (top right), the fan and the gas automatically go to a specific setting `range(0,5,610238:3);out(0,2,610238:2)`, however, when I look at the sliders they both read 0.  Can the slider replicate those settings instead of reading 0?

**Reply by MAKOMO** [reply]
But how did you programmed those events? If you used custom event buttons, you need to configure a corresponding event button "Type" and "Value". Most likely you just added an Action to the default button ON. That will just send the command out, but not log the corresponding event value change.

The correct way to do this, is to define a button called SETUP with the range(0,5,610238:3);out(0,2,610238:2), Type=Air and Value=2. This button can be hidden. Now you note the number of that button (the number all to the left of this row) and set the action of the ON button to "Multiple Event" and put the buttons number in the command field.

Now pressing ON will trigger that (hidden) button which will set the Air event value to 2 and sends out the Phidget commands to the device. Your Air slider will show the actual event value 2.

**Comment by DB-stoutheart** [comment]
Ahhh!!!  Good to know.  Looks like I need to send you more money.  Haha

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by DB-stoutheart** [comment]
Okay making progress....

Another question, with my existing buttons for Gas, I'm using specific voltage values to get the exact numbers I need for specific gas pressures.  I'm doing that by specifying that in the out command, such as `out(0,3.748,610238:3)` which gets me to exactly 4.5kpa.  Is that not being logged then since I don't specify that as a value in the table?  See a picture of my table...

![Screen Shot 2022-02-05 at 2 43 16 PM](https://user-images.githubusercontent.com/94983037/152656602-0fa6fda5-a993-4f31-8b4a-8f69b4bc4259.png)

**Reply by MAKOMO** [reply]
There are several layers here. Artisan events have been designed to work in human-friendly % from 0-100. Your burner valve regulates the gas flow in kPa, while the commands send to this valve from the OUT Phidget are in V. So we have 3 layers. I suggest to work in those user-friendly % event values and let Artisan do the conversion to V under the hood. Who is interested in kPa?

Let's assume your slider definition with the mapping discussed above

<img width="1044" alt="Screenshot 2022-02-06 at 10 27 33" src="https://user-images.githubusercontent.com/13909011/152675419-3a107c51-9df1-4349-946e-2bfe890fa8d9.png">

Note that I added the "%" character as unit. More on this further down.

So we have

```vout(burner%) = 0,042*burner%```

thus we can rearrange this to

```burner% = vout(burner%) / 0,042```

Let's focus on your first button labeled 4.5kpa. Converting this into % by reversing the slider formula as above gives us

```burner%(3.748V) = 3.748V / 0,042 = 89%```

Let's work with those 89% for now thus I suggest in the end to shift this to even 90%.
Below I configured two buttons, the first one is in kpa as you defined it, the second one is in %.

<img width="1044" alt="Screenshot 2022-02-06 at 10 27 33" src="https://user-images.githubusercontent.com/13909011/152675040-2a25891f-445a-46d0-bd76-1bfcae1665a0.png">

Note several things on the % button definition here. First it uses those calculated 89% as event value and sets the event type to Burner. Thus pressing this event button generates an Artisan event which is logged, while the first button is purely emitting a signal to the connected Phidget. Further, the % button sends out a command to the Phidget as the kpa button, but as value it uses the placeholder `{}`which will be replaced by the event value 89 on activation converted via the corresponding Burner sliders factor. Thus the factor/offset as defined for the slider also works for buttons of the same event type! Finally, note that I added the description "4.5kpa". Thus the event generated on activating this button will generate an event with the value 89 and the description "4.5kpa". 

Let's play with this. After pressing just the kpa button we see the following during recording.

<img width="912" alt="Screenshot 2022-02-06 at 10 27 03" src="https://user-images.githubusercontent.com/13909011/152675198-1e7d93aa-44f9-4fb8-9ecb-3c21a0fd59e0.png">

Note that the slider did not move (as you reported) and no event is marked. This is just because no Artisan event was generated. Now let's press that second % button.

<img width="912" alt="Screenshot 2022-02-06 at 10 27 16" src="https://user-images.githubusercontent.com/13909011/152675217-3c587f2f-1941-437c-9004-d750b4c7c0aa.png">

Now an Artisan event was generated and the slider moved to the 89%. The event mark appears in the profile, labeled with 4.5kp using the event description. This is because I ticked "Descr." here in the Event Config.

<img width="1044" alt="Screenshot 2022-02-06 at 10 27 31" src="https://user-images.githubusercontent.com/13909011/152675283-90ad8efa-fbfb-4b95-b012-20351b1927e4.png">

With the "Descr." flag unticked, the event is annotated with the event value and that event unit that I initially set to "%".

<img width="912" alt="Screenshot 2022-02-06 at 10 28 02" src="https://user-images.githubusercontent.com/13909011/152675289-a81f62b7-5501-4865-a38d-7cd3fff8ba2f.png">

You can check the generated event in the event list of the profile under Roast Properties.

<img width="773" alt="Screenshot 2022-02-06 at 10 39 25" src="https://user-images.githubusercontent.com/13909011/152675345-f597ed5a-2f0a-4854-9dee-440d1f579549.png">

Note that now that this event is registered and saved along the profile it can also be replayed if this profile is chosen as a background profile.

**Reply by MAKOMO** [reply]
One more note on this. Most gas valves are non-linear. Mine did not change the flame from turning it down to 50% from 100%, but did significantly change the flame hight from turning it down from 30% to 20%.

If this is the case you might want to tick the Bernoulli flag for your gas slider. Read about Bernoulli Sliders here: 

https://artisan-roasterscope.blogspot.com/2020/10/artisan-v242.html

**Comment by DB-stoutheart** [comment]
So I'm having good luck and bad luck...for some reason, the percentage buttons just are not working properly, I'm not sure if its a build issue or what.  I went through and did the calculations on everything, and run a test with no beans and it worked fine.  I followed that up by clicking the Bernoulli check box on, and recalibrated all the buttons and it was working fine.  I put beans in and ran a roast and everything was wonky, I'd click buttons and the pressure may or may not change or would change very little.  If I used the slider it worked fine though.  I need to step away for the day, but I'm going to come back to this...

Lastly, I need to work out the noise..I'm not sure if my VFD is causing issues, or if I need new probes, but the swings I'm getting are pretty big...at least for my taste.

[SH1228_Sumatra_2022-02-06_1539.pdf](https://github.com/artisan-roaster-scope/artisan/files/8011225/SH1228_Sumatra_2022-02-06_1539.pdf)

Really appreciate all your help...I'm very thankful!

[artisan-settings.txt](https://github.com/artisan-roaster-scope/artisan/files/8011226/artisan-settings.txt)

**Reply by MAKOMO** [reply]
Wow! Another bug. In contrast to the slider actions, those button actions did not respect the event types Bernoulli settings on calculating the values to send out to the module. I just fixed that. Independent of this, your event button % labels should correspond to your event button labels to avoid confusion.

Please download a new "Continuous" build from here:
https://github.com/artisan-roaster-scope/artisan/releases

Yes, your noise is considerable (as far as I can see from your PDF and your heavy smoothing settings; but without being able to look at the raw data in the .alog file). Not sure what this is causing, but it could be the motors or the VFD. Anyhow, your smoothing settings do not make much sense to me. Avoid having curve smoothing set higher than 1 (yours is 5!). Try to put delta_span to 30s and  Delta Smoothing to 0 and test if unticking Polyfit works better for you (for me Polyfit results in a worse result). In your noisy environment you might be better of with RTDs replacing those TCs.

**Reply by DB-stoutheart** [reply]
When I sent you the artisan-settings I had made some adjustments but didn't rename the labels, sorry for the confusion on that. 

In regards to the smoothing, I was playing with different settings.  I need to read a bit more about the settings to understand them. This morning I opened up a previous roast and turned off Polyfit and it already seemed a bit more manageable to me.

I still find it funny that between a Mac and a PC the smoothness of the curve is pretty drastic, even without smoothing.

**Reply by MAKOMO** [reply]
> I still find it funny that between a Mac and a PC the smoothness of the curve is pretty drastic, even without smoothing.

Can you expand on this? I don't understand this fully. Do you say that the smoothness of the same profile looks different when one opens it with Artisan (same version) on a Mac vs a PC? Or do you observe a difference if the profile is recorded using a Mac vs a PC (same Artisan version using the the same equipment)?

**Reply by MAKOMO** [reply]
Regarding your noise issue. Do you run the latest Phidget driver and the latest firmware on your 1048 and the other modules?

**Reply by DB-stoutheart** [reply]
> > I still find it funny that between a Mac and a PC the smoothness of the curve is pretty drastic, even without smoothing.
> 
> 
> 
> Can you expand on this? I don't understand this fully. Do you say that the smoothness of the same profile looks different when one opens it with Artisan (same version) on a Mac vs a PC? Or do you observe a difference if the profile is recorded using a Mac vs a PC (same Artisan version using the the same equipment)?

A month or two ago my PC crashed so I moved Artisan to my Mac (which I prefer anyway) and at that time there was an immediately noticeable difference in the smoothness of the curve. I have my PC back up now, if I have time today I'll try and do a comparison and see if there's still a difference in the curves. 

As for the Phidgets, I checked the firmware about two weeks ago and it was up to date.

**Comment by DB-stoutheart** [comment]
I downloaded the latest continuous build, but its opening as Artisan Viewer only.

**Comment by CarefreeBuzzBuzz** [comment]
You must have another version open somewhere.  Close everything and restart.  
----------------------------------------
Michael Herbert, Head Roaster
Order at www.carefreebuzzbuzz.com <http://www.carefreebuzzbuzz.com/>
Artisan RoasterScope  <https://artisan-scope.org/>Team

**Comment by DB-stoutheart** [comment]
@MAKOMO did a roast last night on my PC, and I was very happy with everything.  I'm replacing my Mac at the end of the week with a new Mac mini so I'm going to wait to retest on that to see how the noise/curve is.  Thank you again for your support and help!

**Comment by MAKOMO** [comment]
Thanks for this feedback. Note that the current version you are using is not even a Release Candidate. I hope to push a Release Candidate out very soon now and hope you will  help us ironing out the last issues in it before releasing the next Artisan version.

**Comment by DB-stoutheart** [comment]
Absolutely happy to help!  The more I learn about the software the better I've become and finding "kinks" in the code.

**Comment by DB-stoutheart** [comment]
Making some progress...Seems I need to tune some more, I expected a bit more consistency between these two back to back roasts.  Just sharing, not complaining. Open to suggestions.
[Comparison.pdf](https://github.com/artisan-roaster-scope/artisan/files/8035879/Comparison.pdf)
